src/systems/physics/CanonicalLinkModelTracker.hh

```diff
@@ -0,0 +1,112 @@
+  /// all of the models that have this canonical link should be updated. It's
+  /// important to preserve topological ordering of the models in case there's
+  /// a nested model that shares the same canonical link (in a case like this,
+  /// the parent model pose needs to be updated before updating the child model
+  /// pose - see the documentation that explains how model pose updates are
+  /// calculated in PhysicsPrivate::UpdateSim to understand why nested model
+  /// poses need to be updated in topological order).
+  ///
+  /// It's possible to loop through all of the models and to update poses if the
+  /// model moved using something like EntityComponentManager::Each, but the
+  /// performance of this approach is worse than using just the moved canonical
+  /// links to determine which model poses should be updated (consider the case
+  /// where there are a lot of non-static models in a world, but only a few move
+  /// frequently - if using EntityComponentManager::Each, we still need to check
+  /// every single non-static model after a physics update to make sure that the
+  /// model did not move. If we instead use the updated canonical link
+  /// information, then we can skip iterating over/checking the models that
+  /// don't need to be updated).
+  class CanonicalLinkModelTracker
+  {
+    /// \brief Save mappings for new models and their canonical links
+    /// \param[in] _ecm EntityComponentManager
+    public: void AddNewModels(const EntityComponentManager &_ecm);
+
+    /// \brief Get a topological ordering of models that have a particular
+    /// canonical link
+    /// \param[in] _canonicalLink The canonical link
+    /// \return The models that have this link as their canonical link, in
+    /// topological order
+    public: std::set<Entity> CanonicalLinkModels(
+                const Entity _canonicalLink) const;
+
+    /// \brief Remove a link from the mapping. This method should be called when
+    /// a link is removed from simulation
+    /// \param[in] _link The link to remove
+    public: void RemoveLink(const Entity &_link);
+
+    /// \brief A mapping of canonical links to the models that have this
+    /// canonical link. The key is the canonical link entity, and the value is
+    /// the model entities that have this canonical link. The models in the
+    /// value are in topological order
+    private: std::unordered_map<Entity, std::set<Entity>> linkModelMap;
+  };
+
+  void CanonicalLinkModelTracker::AddNewModels(
+      const EntityComponentManager &_ecm)
+  {
+    _ecm.EachNew<components::Model, components::ModelCanonicalLink>(
+        [this](const Entity &_model, const components::Model *,
+          const components::ModelCanonicalLink *_canonicalLinkComp)
+        {
+          this->linkModelMap[_canonicalLinkComp->Data()].insert(_model);
+          return true;
+        });
+  }
+
+  std::set<Entity> CanonicalLinkModelTracker::CanonicalLinkModels(
```

Comment:
Is it possible to return a `const &` here so the set doesn't need to be copied?

Reply:
Yep! https://github.com/ignitionrobotics/ign-gazebo/pull/736/commits/11a97df6417c8fde301d63817c81ca816e6df58e

(I had to add an empty entity set member variable to the `CanonicalLinkModelTracker` class to avoid returning references to local/temporary variables)